src/persist-client/src/internal/state.proto

```diff
@@ -60,9 +58,40 @@ message ProtoStateRollup {
     string diff_codec = 5;
     uint64 seqno = 6;
     uint64 last_gc_req = 10;
+    map<uint64, string> rollups = 12;
+
     ProtoTrace trace = 7;
-    repeated ProtoReaderState readers = 8;
-    repeated ProtoWriterState writers = 9;
+    map<string, ProtoReaderState> readers = 8;
+    map<string, ProtoWriterState> writers = 9;
+}
+
+enum ProtoStateFieldDiffType {
+    INSERT = 0;
+    UPDATE = 1;
+    DELETE = 2;
+}
+
+message ProtoStateFieldDiff {
+    bytes key = 1;
+    ProtoStateFieldDiffType diff_type = 2;
+    bytes from = 3;
+    bytes to = 4;
+}
+
+message ProtoStateDiff {
+    string applier_version = 1;
+
+    uint64 seqno_from = 2;
+    uint64 seqno_to = 3;
+    string latest_rollup_key = 4;
+
+    // WIP columnar encoding of these just in case?
```

Comment:
still haven't gotten around to prototyping this. worst case, we can change it right before oct 11